### ISSUE TITLE
How to work around RDT flaws on skylake servers?

### ISSUE THREAD

**Issue by vatiminxuyu** [intel/intel-cmt-cat#95]
"RDT: Disable most RDT features on Skylake server"[1] states that there are serious errata in CMT, MBM and CAT sub-features of resource director technology. 

I also reproduced some fatal flaws of RDT in my server, such as SKZ4, SKZ14, SKZ15 [2], which directly affected the judgment of data accuracy.

For example, when there are multiple VMs,
1) the sum of the llc occupancy of each VM obtained by CMT may far exceed the physical llc size; 
2) when I isolate the llc ways of one VM with CAT, the llc occupancy obtained by CMT may also  far exceed the theoretical value. And I can't know if there is a problem with CMT or a problem with CAT.
3) open issue: 65% - 94% mba cos value generates the same result 90% [3]

If I want to take advantages of RDT on skylake servers, how to work around the problems above?

[1] https://bugs.launchpad.net/intel/+bug/1713619
[2] https://www.intel.com/content/dam/www/public/us/en/documents/specification-updates/6th-gen-x-series-spec-update.pdf
[3] https://github.com/intel/intel-cmt-cat/issues/69

**Comment by kmabbasi** [comment]
Hi Xu Yu,

Thanks for your feedback.

From our testing on Skylake servers, we haven't seen cases where CMT reports occupancy significantly higher than expected values. This could be workload dependent.

Some suggestions to improve accuracy could be:

1. Ensure each VM is running on a single socket. If VM cores extend across multiple sockets LLC occupancy could exceed physical LLC size.
2. It is difficult to determine the cause of inaccuracy. Please ensure that any VM data already in the LLC is flushed after changing CAT configuration as this will still be counted. Please see issue #62 for more detail around this.
3. You can try to use MBA SW controller feature added to rdtset app in the latest release. This allows to use SW controller to limit local memory B/W to specified MBps. This can help to provide more deterministic B/W.

Thanks,
Khawar

**Comment by vatiminxuyu** [comment]
Hi Khawar,

Thanks so much for your kind and effective suggestions. And there is another confusing question, described with the following experiment.

In my experiment, there are several VMs on a single socket, each with exclusive cores (i.e., vcpupin with non-overlapping cpus), and runs cache/memory intensive benchmarks.

1. First, I assign each VM a different RMID, in descending order of available RMIDs, i.e., 191, 190, 189, ..., etc.
The sum of the LLC occupancy of all VMs obtained by CMT has risen from 33M (physical LLC size) to ~50M in hours (or days, it depends).

2. Then, I reassign each VM a different RMID, in ascending order of available RMIDs, i.e., 1, 2, 3, ..., etc. 
The sum of the LLC occupancy of all VMs _immediately_ drops to 33M. BUT. Hours (or days, it depends) later, the sum has also risen up to ~50M.

3. Then, I reassign each VM a different RMID, in descending order of available RMIDs, i.e., 191, 190, 189, ..., etc.
I suppose that the sum would be 33M. BUT, the sum is ~50M actually.


Besides the known issue that CMT on Skylake may over count, the caches tagged with RMID 191, 190, 189, ..., etc in round-1 should be evicted and restored with RMID 1, 2, 3, ... etc in round-2, and the sum llc occupancy should be 33M when round-3 begins.

Any idea on this experiment?


Thanks,
Yu.

**Comment by kmabbasi** [comment]
Hi Yu,

Thanks for investigating this issue.

We will try to reproduce this behavior in our system. We will get back to you once we have update.

Thanks,
Khawar

**Comment by vatiminxuyu** [comment]
Hi Khawar,

Any updates?

Thanks,
Yu.

**Comment by kmabbasi** [comment]
Hi Yu,

Sorry for late response! 

We tried to re-produce issue. For that we used pqos tool (HW interface) and "stress-ng" utility for putting loads on 20 cores with matrix size 3072. All 20 cores have different RMIDs. The tests had been running for ~21 hours but we didn't find any issue.

If you would like to provide more information which can help us to reproduce the issue, we will be happy to investigate further.

Thanks,
Khawar
